offline/cable_cru_TRENDY.F90

```diff
@@ -408,38 +551,80 @@ SUBROUTINE cru_get_daily_met(CRU, LastDayOfYear)
   ! opening and closing the io stream unnecessarily, but I think it's a minor
   ! evil.
 
-  ! Address prevTmax first
+  ! There's a serious (?) question to be answered here- what do we do about leap
+  ! years? There are two ideas which are at odds with eachother:
+  ! 1. We want to retrieve that correct data for the relevant year used for
+  !   recycling, which implies leap years should be based on the met recycling
+  !   year. This prevents instances where we try to access the 366th day of a
+  !   year with 365 days, or only get 365 days from a 366 day year.
+  ! 2. We often want to perform yearly averages or accumulations over a year.
+  !   These accumulations/averages occur elsewhere in the code, and thus use the
+  !   number of days from the core simulation year. This imples leap years
+  !   should be based on the core simulation year, to prevent instances where
+  !   we accumulate meteorology over 366 days but average over 365 days and vice
+  !   versa.
+  ! Address previous variables first
   ! Assume it's a regular day of the year
   DummyYear = CRU%cYear
   DummyDay = CRU%CTStep - 1
 
   ! Special handling at the first day of the year
-  IF (CRU%CTSTEP == 1) THEN
+  ! We need to be very careful with start of year handling due to leap years.
+  ! If we do things in the incorrect order, we can end up with asking for the
+  ! 366th day of a non-leap year
+  IF (CRU%CTStep == 1) THEN
     ! Go back to previous year
     DummyYear = CRU%cYear - 1
-    DummyDay = 365
-    IF ((CRU%LeapYears) .AND. (is_leapyear(DummyYear))) THEN
-      DummyDay = 366
-    END IF
   END IF
 
-  ! Was the Tmax recycled?
+  ! Now we have to handle each of the "prev" variables individually
+  ! Tmax
   IF (CRU%isRecycled(Tmax)) THEN
-    DummyYear = CRU%RecycStartYear + MOD(DummyYear - 1501, CRU%metRecyc)
+    DummyYear = CRU%RecycStartYear + MOD(DummyYear - 1501, CRU%MetRecycPeriod)
   END IF
 
+  DummyDay = 365
+  IF ((CRU%LeapYears) .AND. (is_leapyear(DummyYear))) THEN
+    DummyDay = 366
+  END IF
   ! Now we just need to call cru_read_metvals with the Tmax Dataset reader and
   ! the prevTmax array to write to
   CALL read_metvals(CRU%MetDatasets(Tmax), CRU%Met(prevTmax)%MetVals, land_x,&
     land_y, DummyYear, DummyDay, CRU%LeapYears, CRU%xDimSize, CRU%yDimSize,&
     CRU%DirectRead)
 
-  ! Now do nextTmin
+  ! vph0900
+  DummyYear = CRU%cYear
+  DummyDay = CRU%CTStep - 1
+
+  IF (CRU%CTStep == 1) THEN
+    ! Go back to previous year
+    DummyYear = CRU%cYear - 1
+  END IF
+  
+  IF (CRU%isRecycled(vph0900)) THEN
+    DummyYear = CRU%RecycStartYear + MOD(DummyYear - 1501, CRU%MetRecycPeriod)
+  END IF
+
+  DummyDay = 365
+  IF ((CRU%LeapYears) .AND. (is_leapyear(DummyYear))) THEN
+    DummyDay = 366
+  END IF
+
+  ! Now we just need to call cru_read_metvals with the Tmax Dataset reader and
+  ! the prevTmax array to write to
+  CALL read_metvals(CRU%MetDatasets(vph1500), CRU%Met(prevvph1500)%MetVals, land_x,&
```

Comment:
Why do we check "isRecycled(vph**0900**)" but use vph**1500** in `read_metvals`?

I am also wondering if it would be worth having a function to "setDummyTime_and_ReadTheData" since it's always done in the same way and it's only the variable indexes that change. Would need either 2 functions or a logical to swap between previous time and next time calculations.

Such a function would avoid the mistake at the start of this comment, if it is a mistake.

Reply:
I think it might be better if I just pull the date calculation into a separate routine? Something like
```
SUBROUTINE get_met_date(SimYear, SimDay, IsRecycled, RecycleStart, RecyclePeriod, LeapYears, MetYear, MetDay)
! Some routine that modifies MetYear and MetDate based on the recycling parameters
END SUBROUTINE get_met_date
```
Then there's a workflow that works identically for any variable- compute the date, then read the data. We also get a nice division of responsibility that way- one function gets the date, one function gets the data.